Extend SPSP

angriman opened the pull request:
This PR provides a feature requested in our mailing list, i.e., to extend SPSP so that the user can specify target nodes. In particular:
- SPSP has now another constructor and a function to set target nodes
- In the case of target nodes, SPSP uses new `MultiTargetBFS/Dijkstra` algorithms that stop the exploration once all target nodes have been visited -- I preferred not implementing this within `BFS` and `Dijkstra` because `SSSP` would then implement too many functions relative to different features and I find this confusing for the users.

Additional changes:
- Member variables that store the distances were moved from derived classes to the base class `STSP` so that we can "un-virtualize" `STSP` functions.
- Parametrization of the tests of the `distance` module and fix of bugged tests.
- Python tests for the `distance` module are moved to another file.
- Minor fixes for the documentation of `BFS`.

fabratu commented on networkit/cpp/distance/test/DistanceGTest.cpp:
```diff
@@ -160,41 +178,31 @@ TEST_F(DistanceGTest, testBidirectionalBFS) {
-        G1.setWeight(u, v, Aux::Random::probability());
-    });
+    auto G = generateERGraph(500, isDirected() ? 0.05 : 0.02);
+    const auto infDist = std::numeric_limits<edgeweight>::max();
```
Only minor: the value of `infDist` is used for several tests. Maybe it is a good idea to change it to be accessible by the complete class.

fabratu reviewed (changes requested):
Hm, a general comment (likely outside the scope of this PR): I think the way the shortest path algorithms are organized can be improved. `SSSP` and `STSP` are abstract classes and cannot be used directly, but `SPSP` and `APSP` can. From a user perspective this can lead to confusion, especially if one is not familiar with inheritance and abstract classes. 

Another thing is the naming scheme with respect to the current feature set. From the perspective of a inexperienced user, it is not so clear what interface to use. Do I use `SPSP` or a child of `STSP`; what exactly differentiates `SSSP` and `STSP` (due to the same interface)?

If possible, one way around it would be to hide the abstract base classes in Python interface (can look into it) or document properly what subclasses exist (for example when spawning a `SSSP` and `STSP` object).

angriman commented on networkit/cpp/distance/MultiTargetBFS.cpp:
```diff
@@ -0,0 +1,66 @@
+        const auto it = targetsSet.find(u);
+        if (it != targetsSet.end()) {
+            targetIdx.emplace(u, distances.size());
+            distances.push_back(static_cast<edgeweight>(level));
```
I agree, but this might not be possible in C++ (e.g., when the targets are stored in an `unordered_set`).

fabratu replied:
Right, I wasn't aware of `unordered_set`. In that case I would suggest to document the behavior for C++ (the need to iterate through `getTargetIndexMap()` to get correct target value). For the Python interface I think it is a better solution to return a correctly indexed distance list (since Cython mapping is `list` <-> `vector`).

angriman replied:
This is now implemented by keeping a copy of the targets list, which is used to sort the distances.

angriman commented on include/networkit/distance/AStarGeneral.hpp:
```diff
@@ -67,13 +67,14 @@ class AStarGeneral final : public STSP {
         do {
             top = heap.extract_top();
             if (top == target) {
+                distance = distFromSource[target];
```
Good catch, now `distance` is set to infinity by default at the beginning of the algorithm.

angriman commented:
Thanks for the review. I agree that distance algorithms need to be restructured and simplified; since this PR brings a new feature, we should do that separately. I addressed all the comments.

fabratu commented on networkit/distance.pyx:
```diff
@@ -32,9 +32,13 @@ cdef extern from "cython_helper.h":
 cdef extern from "<networkit/distance/STSP.hpp>":
 	cdef cppclass _STSP "NetworKit::STSP"(_Algorithm):
 		_STSP(_Graph G, node source, node target, bool_t storePred) except +
```
This doesn't seem to be necessary, but I would suggest to also add the second constructor for `STSP`.

angriman replied:
Unfortunately, it does not seem to be possible to have a template constructor in Cython. Let me know if I am wrong.

fabratu replied:
Hm, I was just thinking to have the same as for `SPSP` (no general template version): 

```
_STSP(_Graph G, node source, vector[node].iterator targetsFirst, vector[node].iterator targetsLast, bool_t) except +
```

But it's really only for consistency with the other base classes.

angriman replied:
Ah, I see. I added it.

fabratu commented on networkit/cpp/distance/BidirectionalBFS.cpp:
```diff
@@ -15,7 +15,7 @@ void BidirectionalBFS::run() {
         WARN("Treating the graph as unweighted!");
 
-    stDist = 0;
+    distance = 0;
```
If BidirectionalBFS is constructed using the new iterator-based target-list constructor, the algorithm would crash (due to `target=none`)?

angriman replied:
Good point, now both `BidirectionalBFS/Dijkstra` have their own constructor that takes a single target node.

fabratu commented on include/networkit/distance/STSP.hpp:
```diff
@@ -69,17 +90,79 @@ class STSP : public Algorithm {
+    template <class InputIt>
+    void setTargets(InputIt targetsFirst, InputIt targetsLast) {
+        assert(std::all_of(targetsFirst, targetsLast, [&](auto u) { return G->hasNode(u); }));
+        targets.assign(targetsFirst, targetsLast);
```
Should `target` be set back to `none`? Or otherwise should we drop variable `target` and use `targets`-vector in all cases (with implied changes for BidirectionalBFS/Dijkstra)?

angriman replied:
I think that none of the possibilities is satisfactory. It's weird for a class dealing with a single target to have it stored in a vector, while `target` is never used by `MultiTargetBFS/Dijkstra` -- so there is no point in setting it to `none`. I think that this issue can be solved by restructuring these algorithms.

fabratu replied:
Agreed.

fabratu commented on networkit/cpp/distance/BidirectionalDijkstra.cpp:
```diff
@@ -15,7 +15,7 @@ void BidirectionalDijkstra::run() {
         WARN("The graph is unweighted, use BidirectionalBFS for better "
              "efficiency!");
-    stDist = 0.;
+    distance = 0.;
```
Same as for `BidirectionalBFS`.

fabratu reviewed (dismissed):
LGTM.

fabratu reviewed (approved):
LGTM.

angriman commented:
Merging now because this PR has been approved for more than 7 days before I force-pushed to resolve a trivial conflict.
